### DISCUSSION TITLE
Why not just use npm instead of your own manager?

### DISCUSSION THREAD

**Discussion by Psychosynthesis** [pulsar-edit/.github#211] [General]
Why do you even need your own built-in ppm package manager?

What specific problems does it solve that can't be solved with npm?

I'm interested because I want to get involved in editor development (I'm frustrated with the decline of Atom) and one of the most disgusting things that annoys me is the complex multi-stage bootstrapping of the project (I'm judging by the latest version of Atom, but it seems that you have about the same situation). The more custom tools are used, the more complex the code as a whole and, as a result, worse.

Npm is a great tool, almost everyone who works with the web today has it, so why not use it? It would eliminate at least one step when deploying a project.

**Comment by Daeraxa** [comment]
Probably just a case of "if it ain't broke, don't fix it". ppm is literally just apm with some tweaks. Ultimately it is still npm of a version that we know works with the editor. Are you saying we should be using the machine installed version of npm to manage Pulsar packages? Or that we should just be shipping npm with Pulsar (which is essentially what we are doing anyway)?

> one of the most disgusting things that annoys me is the complex multi-stage bootstrapping of the project (I'm judging by the latest version of Atom, but it seems that you have about the same situation).

Pulsar is built in much the same way of most Electron projects. We don't have the old bootstrapping, we chucked that all in the bin pretty much on day 1. All you need is to initialise a submodule for ppm then a `yarn install` followed by a `yarn build` for Pulsar then ppm. It isn't a particularly complex set of build instructions. (Oh and in case you were wondering, "why Yarn and not npm?" That goes back to the first days as well where Yarn was literally the only tool that would work to build Pulsar).

**Comment by confused-Techie** [comment]
While I disagree that complexity = worse code, there are a few reasons why PPM is still being used.

1. As you know it's already built into Atom, meaning it's already a part of the codebase, and with our limited team and many other things to focus on, removing it has never been a huge priority, although a few of us have tried a few times to move PPM into the core of the editor to make things simpler. While it hasn't seen activity in a while, I actually have a draft PR where I attempt to do exactly this: [`pulsar-edit/pulsar#445`](https://github.com/pulsar-edit/pulsar/pull/445).
2. While you are right, that PPM is similar to NPM, and in fact PPM uses NPM under the hood, it doesn't do all of the same things. For example a package published to NPM needs to write in the scripts to translate things like LESS into CSS, whereas that's never been a requirement with PPM. That's because PPM will automatically transpile things like Less into CSS, whereas NPM won't automatically do this. Plus PPM will install packages into the custom `~.pulsar` directory, etc. So while similar these two tools do have different tasks they need to do
3. The existing community packages. As you may, or may not know, the Pulsar team managed to archive every single community package that was published to Atom, before they shut everything down. If we suddenly just use NPM, there would be no way to publish all of these packages. Every single community package would just be lost. Sure NPM could download these from GitHub, but how would they be distributed? Atom had it's own custom package registry, and as such so does Pulsar to keep things working. If we remove the one tool that interacts with this custom registry, then it becomes useless and we lose those packages.
4. Finally, while you are right many developers may already have NPM, you do have to remember that NPM is really only used in a subset of developers, those working with NodeJS and JavaScript. You'd be surprised how many Pulsar users don't actually have NPM installed, because say they develop in C++, or just use Pulsar for writing Markdown, or develop in Python, etc. 

---

With all of this said, there would be a lot of problems to solve by removing PPM, for not many benefits, the code is already written, it was already an expected part of Atom, and if anything I'd say our longer terms goals will likely be to simplify PPM and bring it into the editor, rather than remove it, as we haven't seen the need to delete it, since really the only complexity it brings to end users, (who aren't just installing Pulsar from our Regular or Rolling releases, since of course it's already built there) is about 3 terminal commands:

```bash
git submodule init
git submodule update
yarn run build:apm
```

**Comment by mauricioszabo** [comment]
@Psychosynthesis PPM is exactly the same package manager that Atom used to have - we just renamed the project and bumped the node version, that's all.

The "multi step" that you mention was removed from Atom - what we do have is a "install" then "rebuild packages for electron". That's all - the rest is for ppm, and it's not actually _needed_ to run the editor.
